redirect to login page if unauthorized 401

jaybe78 opened the pull request:
## Context (Bugs #470  #482)
When query response returns 401, the user must be redirected to login page

## Tasks done
I've seen some redundant code in the hooks that fetch data and deal with issues.
So I created a common hook that deal with 401 and other errors:

```
const useQueryError = () => {
    const queryClient = useQueryClient();
    const [errorResponse, setErrorResponse] = useState<any>(null);
    const {
        errorLogs: { showError },
    } = useApp();
    useEffect(() => {
        (async function doIfError() {
            if (errorResponse) {
                // @ts-ignore
                const { statusCode } = errorResponse;
                if (statusCode === 401) {
                    await queryClient.invalidateQueries('health');
                } else {
                    const { message } = errorResponse;
                    const [first, ...rest] = message.split('\n');
                    showError({ title: first, body: rest.join('\n') });
                }
            }
        })();
    }, [errorResponse]);
    return [errorResponse, setErrorResponse];
};
```

Then for each `useQuery` I import the previous hook and set the response in the `onError` callback as shown below

```
 const [, setErrorResponse] = useQueryError();

 return useQuery<ApiStatusResults, ApiError>({
       ...,
        onError: (result) => setErrorResponse(result.error),
    });
```

There was also an issue with the popup showing up for a split second before the redirection actually occurs.
That was because of a piece of code in `ExploreSideBar.tsx` that i've commented out

```
if (exploresResult.status === 'error') {
        console.log('exploresResult sidebar')
        onBack();
        const [title, ...lines] =
            exploresResult.error.error.message.split('\n');
        showError({ title, body: lines.join('\n') });
        return null;
    }
```

The above code is no longer needed as the redirection is done by the new hook `useQueryError`

ZeRego commented:
@jaybe78 @owlas 
For some context, in the app we have 2 main ways to display errors:
- **in a toast** This is the default way to show errors
- **in the error drawer** We use this to show connection & compile errors since they have big messages and we wanted them to be persistent and easier to read and copy.

```
    const {
        showToastError, // show in a toast
        errorLogs: { showError }, // show in the drawer
    } = useApp();
```

I like how you made a util hook so we could use it everywhere but is going to display the errors in the drawer. And we are already calling a function to show the error. 

Proposal: 

In AppProvider or as a separate hook, have a single function to handle errors that we can use everywhere. 
This hook will decide how to handle the error based on the error code and type. 

```
// pseudo code
showError = (error) => {
    const { statusCode, message, type } = error;
    if (statusCode === 401) {
        await queryClient.invalidateQueries('health');
    } else {
        const [first, ...rest] = message.split('\n');
        
        if(type === 'CompileError' || type === 'ConnectionError') {
                showError({ title: first, body: rest.join('\n') }); 
        } else {
                showToastError({ title: first, subtitle: rest.join('\n') });
        }
    }
}
```

This is a similar approach to what you have now and we can consolidate the way we handle errors.

jaybe78 commented:
Thanks @ZeRego 
I developed that custom hook because I've seen in multiple places the same repeated code below:
```
useEffect(() => {
        if (query.error) {
            const [first, ...rest] = query.error.error.message.split('\n');
            showError({ title: first, body: rest.join('\n') });
        }
    }, [query.error, showError]);
```
So, I might be missing something but those issues are always displayed in the drawer ?

Can't we reuse the hook I've already developed and modify a bit, with what you wrote, so that it handles both types:

```
const useQueryError = () => {
    const queryClient = useQueryClient();
    const [errorResponse, setErrorResponse] = useState<any>(null);
    const {
        errorLogs: { showError },
    } = useApp();
    useEffect(() => {
        (async function doIfError() {
            if (errorResponse) {
                // @ts-ignore
                const { statusCode } = errorResponse;
                if (statusCode === 401) {
                    await queryClient.invalidateQueries('health');
                } else {
                    const { message } = errorResponse;
                    const [first, ...rest] = message.split('\n');
                    if(type === 'CompileError' || type === 'ConnectionError') {
                       showError({ title: first, body: rest.join('\n') }); 
                    } else {
                        showToastError({ title: first, subtitle: rest.join('\n') });
                    }
                }
            }
        })();
    }, [errorResponse]);
    return [errorResponse, setErrorResponse];
};
```
We just need to send the error `type` in the `onError` callback
```
 return useQuery<ApiStatusResults, ApiError>({
       ...,
        onError: (result) => setErrorResponse(result.error, errorType),
    });
```

ZeRego commented:
@owlas can you confirm what are the error types we want to display in the error drawer ?

owlas commented:
@ZeRego it's all errors during compilation and running queries, there are a few types, including a generic 500. 

Not sure that's the way we can split what goes into the drawer vs. a modal.

Most errors in the explore view should go in the drawer I think

ZeRego commented:
That makes it harder to keep our code generic.
If we want that behaviour, the best way is for `useQueryError` to have the argument `showInDrawer?:boolean` 
and show the error based on that instead of the error type.

ZeRego commented on packages/frontend/src/hooks/useQueryError.ts:
```diff
@@ -0,0 +1,30 @@
+
+const useQueryError = () => {
+    const queryClient = useQueryClient();
+    const [errorResponse, setErrorResponse] = useState<any>(null);
```
We could type it with `ApiError` and remove the ts-ignore in line 14

jaybe78 replied:
done

jaybe78 commented on packages/frontend/src/components/ExploreSideBar.tsx:
```diff
@@ -217,13 +217,14 @@ export const ExplorePanel = ({ onBack }: ExplorePanelProps) => {
     }
-    if (exploresResult.status === 'error') {
+    /* if (exploresResult.status === 'error') {
+        console.log('exploresResult sidebar')
         onBack();
         const [title, ...lines] =
             exploresResult.error.error.message.split('\n');
         showError({ title, body: lines.join('\n') });
         return null;
-    }
+    } */
```
do we need `onBack()` ?

jaybe78 replied:
done

ZeRego commented on packages/frontend/src/App.tsx:
```diff
@@ -33,6 +33,13 @@ const queryClient = new QueryClient({
         queries: {
             refetchOnWindowFocus: false,
+            onError: async (result) => {
+                // @ts-ignore
+                const { error: { statusCode } = {} } = result;
+                if (statusCode === 401) {
```
could we avoid this ts-ignore if we did something like this `if(result.error?.statusCode === 401)` ?

jaybe78 replied:
done

ZeRego reviewed (approved):
LGTM 🎉

ZeRego commented:
Closes: #482
